### PR TITLE
Reclaim StarPU tiles on mark_output(false) for incremental sessions

### PR DESCRIPTION
**Pull request by Muxas**
## Summary

Multi-epoch `torch_nntile` training (e.g. `train_deep_relu_mnist.py`) kept growing StarPU memory because the incremental session never reclaimed buffers for tensors whose Python bindings had cleared `mark_output`.

This change keeps a **single incremental** `TensorGraph` / `TileGraph` session and uses `mark_output(false)` as the reclaim signal:

1. **`Runtime::compile()`** refreshes tile `is_output` from logical `TensorNode` marks (sets `is_input` only when logical is marked input; never clears tile input marks — torch_nntile drives persistence via `mark_output`).
2. **`invalidate_non_output_tiles()`** walks allocated tiles and `invalidate_submit`s those that are neither input nor output **and** not used by the pending (not-yet-executed) op slice (`executed_op_end_` watermark).
3. **`allocate_missing_tiles()`** skips reallocating unmarked historical tiles that DCE still lists as live unless a pending op needs them.
4. **`append_tensor_graph_phase`** refreshes marks on reused tiles.
5. **`train_full_batch_step`** drops `logits` / `loss` after run + host readout so bindings clear `mark_output` before the next compile.

Also: without `--compare-torch`, `train_deep_relu_mnist.py` no longer reprints the final nntile loss table (per-epoch `[nntile]` lines already cover that).

## Verification

- C++: `test_append_tensor_graph_phase`, `test_mlp`, `test_tile_graph` — pass (includes new reclaim test).
- Python: `torch_nntile/tests/test_graph_execution.py` — 14/14 pass (torch **2.9.1**).
- RSS probe (DeepReLU MNIST-like, 8 epochs after warmup): growth ≈ **60 KB** (was unbounded before).

## Docs

- [`docs/dev/torch_nntile_tensor_architecture.md`](docs/dev/torch_nntile_tensor_architecture.md) — incremental session reclaim rule.
- [`torch_nntile/README.md`](torch_nntile/README.md) — lifetime guidance.

<div><a href="https://cursor.com/agents/bc-9386afd5-a6d0-425e-aad4-f833124cf1b5"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-9386afd5-a6d0-425e-aad4-f833124cf1b5"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Comment by Muxas** [comment]
@cursor review
